File: data/repositories/appstream_repository.py
import gi
gi.require_version('AppStream', '1.0')
from gi.repository import AppStream
from typing import List
import json
from pathlib import Path
from domain.entities import Package
from domain.repositories import FeaturedRepository

class AppStreamFeaturedRepository(FeaturedRepository):
    def __init__(self, alpm_repo, flatpak_repo):
        self.alpm_repo = alpm_repo
        self.flatpak_repo = flatpak_repo
        self.pool = AppStream.Pool()
        self.pool.load()
        self._pkgname_map = None
        
    def get_packages_from_file(self, config_filename: str, fallback_ids: List[str]) -> List[Package]:
        config_path = Path(__file__).parent.parent.parent / "data" / config_filename
        app_ids = fallback_ids
        
        if config_path.exists():
            try:
                with open(config_path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    if isinstance(data, list):
                        app_ids = data
            except Exception as e:
                print(f"Error loading {config_filename}: {e}")
                
        results = []
        for app_id in app_ids:
            try:
                comp_box = self.pool.get_components_by_id(app_id)
                comps = comp_box.as_array() if hasattr(comp_box, 'as_array') else list(comp_box)
                if not comps:
                    # Fallback: Treat app_id as a pacman or AUR package name directly
                    installed = self.alpm_repo.is_installed(app_id)
                    installed_version = self.alpm_repo.get_installed_version(app_id) if installed else ""
                    
                    if installed:
                        pkg_type = "pacman"
                        version = installed_version
                        desc = "Pacote de sistema instalado."
                    else:
                        version = ""
                        desc = ""
                        for db in self.alpm_repo.syncdbs:
                            try:
                                pkg = db.get_pkg(app_id)
                                if pkg:
                                    version = pkg.version
                                    desc = pkg.desc or ""
                                    break
                            except Exception:
                                pass
                        
                        if version:
                            pkg_type = "pacman"
                        else:
                            # Not in pacman repos, assume AUR
                            pkg_type = "aur"
                            version = "latest"
                            desc = "Ferramenta/Biblioteca de jogos (AUR)"
                            
                    results.append(Package(
                        name=app_id,
                        title=app_id.replace("-bin", "").replace("-git", "").replace("-", " ").title(),
                        desc=desc or "Pacote utilitário de jogos.",
                        version=version or "latest",
                        type=pkg_type,
                        installed=installed,
                        installed_version=installed_version,
                        icon="package"
                    ))
                    continue
                
                # Prefer the component that has a package name
                selected_comp = comps[0]
                for c in comps:
                    if c.get_pkgname():
                        selected_comp = c
                        break
                
                pkg_name = selected_comp.get_pkgname()
                
                if pkg_name:
                    pkg_type = "pacman"
                    installed = self.alpm_repo.is_installed(pkg_name)
                    installed_version = self.alpm_repo.get_installed_version(pkg_name) if installed else ""
                    
                    version = installed_version
                    if not version:
                        for db in self.alpm_repo.syncdbs:
                            try:
                                pkg = db.get_pkg(pkg_name)
                                if pkg:
                                    version = pkg.version
                                    break
                            except Exception:
                                pass
                    if not version:
                        version = "latest"
                else:
                    pkg_name = selected_comp.get_id()
                    pkg_type = "flatpak"
                    installed = self.flatpak_repo.is_installed(pkg_name)
                    installed_version = ""
                    if installed:
                        installed_flatpaks = self.flatpak_repo._get_installed_flatpaks()
                        installed_version = installed_flatpaks.get(pkg_name, "")
                    version = installed_version or "latest"
                
                icon_name = ""
                icons = selected_comp.get_icons()
                if icons:
                    selected_icon = icons[0]
                    for ic in icons:
                        if hasattr(ic, 'get_kind') and ic.get_kind() == 1:
                            selected_icon = ic
                            break
                    if hasattr(selected_icon, 'get_name'):
                        icon_name = selected_icon.get_name()
                        if icon_name.endswith((".png", ".svg", ".xpm")):
                            icon_name = icon_name.rsplit(".", 1)[0]
                
                if not icon_name:
                    icon_name = pkg_name or app_id
                
                results.append(Package(
                    name=pkg_name,
                    title=selected_comp.get_name() or pkg_name,
                    desc=selected_comp.get_summary() or "",
                    version=version,
                    type=pkg_type,
                    installed=installed,
                    installed_version=installed_version,
                    icon=icon_name
                ))
            except Exception as e:
                print(f"Error processing app_id {app_id}: {e}")
                
        return results

    def get_featured_packages(self) -> List[Package]:
        fallback_ids = [
            "org.mozilla.firefox",
            "com.visualstudio.code",
            "org.kde.kdenlive",
            "org.gimp.GIMP",
            "org.videolan.VLC",
            "org.inkscape.Inkscape",
            "org.blender.Blender",
            "com.valvesoftware.Steam",
            "org.libreoffice.LibreOffice",
            "org.telegram.desktop"
        ]
        return self.get_packages_from_file("featured_apps.json", fallback_ids)

    def get_gaming_packages(self) -> List[Package]:
        fallback_ids = [
            "com.valvesoftware.Steam",
            "net.lutris.Lutris",
            "com.github.benjamimgois.goverlay",
            "gamemode",
            "mangohud",
            "com.discordapp.Discord",
            "com.heroicgameslauncher.hgl",
            "hydra-games-launcher-bin",
            "wine",
            "winetricks",
            "com.github.Mathew148.Protontricks",
            "net.davidotek.pupgui2"
        ]
        return self.get_packages_from_file("gaming_apps.json", fallback_ids)

    def get_hero_apps(self) -> List[dict]:
        """Return up to 6 apps with hero image data for the Explore page carousel.

        Each dict contains:
          - All Package fields (name, title, desc, version, type, installed, icon)
          - hero_image: URL of the first DEFAULT/SOURCE screenshot
          - brand_color_light: branding primary color for light scheme (or default)
          - brand_color_dark:  branding primary color for dark scheme (or default)
        """
        hero_ids = [
            "com.valvesoftware.Steam",
            "org.libreoffice.LibreOffice",
            "net.lutris.Lutris",
            "com.heroicgameslauncher.hgl",
            "com.visualstudio.code",
            "org.gimp.GIMP",
        ]

        # Fallback accent colors (Catppuccin Mocha tones) per app when branding absent
        fallback_colors = {
            "com.valvesoftware.Steam":        {"light": "#c6d0f5", "dark": "#1b2838"},
            "org.libreoffice.LibreOffice":    {"light": "#c6d0f5", "dark": "#18A303"},
            "net.lutris.Lutris":              {"light": "#f4b8e4", "dark": "#cc5500"},
            "com.heroicgameslauncher.hgl":    {"light": "#ac9aac", "dark": "#1c202c"},
            "com.visualstudio.code":          {"light": "#c6d0f5", "dark": "#0078d4"},
            "org.gimp.GIMP":                  {"light": "#f4b8e4", "dark": "#4e3e6e"},
        }

        IMAGE_KIND_SOURCE = AppStream.ImageKind.SOURCE
        results = []

        for app_id in hero_ids:
            try:
                comp_box = self.pool.get_components_by_id(app_id)
                comps = comp_box.as_array() if hasattr(comp_box, 'as_array') else list(comp_box)
                if not comps:
                    continue
                comp = comps[0]

                # ── Hero image: first DEFAULT screenshot, SOURCE size ──────────────
                hero_image = ""
                for sc in comp.get_screenshots_all():
                    for img in sc.get_images_all():
                        if img.get_kind() == IMAGE_KIND_SOURCE:
                            hero_image = img.get_url() or ""
                            break
                    if hero_image:
                        break

                # ── Branding colors ───────────────────────────────────────────────
                fb = fallback_colors.get(app_id, {"light": "#c6d0f5", "dark": "#1e1e2e"})
                brand_light = fb["light"]
                brand_dark  = fb["dark"]
                branding = comp.get_branding()
                if branding:
                    try:
                        c = branding.get_color(AppStream.ColorKind.PRIMARY, AppStream.ColorSchemeKind.LIGHT)
                        if c:
                            brand_light = c
                    except Exception:
                        pass
                    try:
                        c = branding.get_color(AppStream.ColorKind.PRIMARY, AppStream.ColorSchemeKind.DARK)
                        if c:
                            brand_dark = c
                    except Exception:
                        pass

                # ── Package metadata ──────────────────────────────────────────────
                pkg_name = comp.get_pkgname() or app_id
                if comp.get_pkgname():
                    pkg_type = "pacman"
                    installed = self.alpm_repo.is_installed(pkg_name)
                    installed_version = self.alpm_repo.get_installed_version(pkg_name) if installed else ""
                    version = installed_version
                    if not version:
                        for db in self.alpm_repo.syncdbs:
                            try:
                                pkg = db.get_pkg(pkg_name)
                                if pkg:
                                    version = pkg.version
                                    break
                            except Exception:
                                pass
                    version = version or "latest"
                else:
                    pkg_name = app_id
                    pkg_type = "flatpak"
                    installed = self.flatpak_repo.is_installed(pkg_name)
                    installed_version = ""
                    if installed:
                        installed_flatpaks = self.flatpak_repo._get_installed_flatpaks()
                        installed_version = installed_flatpaks.get(pkg_name, "")
                    version = installed_version or "latest"

                # ── Icon ──────────────────────────────────────────────────────────
                icon_name = ""
                icons = comp.get_icons()
                if icons:
                    icon_name = icons[0].get_name() if hasattr(icons[0], 'get_name') else ""
                    if icon_name.endswith((".png", ".svg", ".xpm")):
                        icon_name = icon_name.rsplit(".", 1)[0]
                if not icon_name:
                    icon_name = pkg_name

                results.append({
                    "name":              pkg_name,
                    "title":             comp.get_name() or pkg_name,
                    "desc":              comp.get_summary() or "",
                    "version":           version,
                    "type":              pkg_type,
                    "installed":         installed,
                    "installed_version": installed_version,
                    "icon":              icon_name,
                    "hero_image":        hero_image,
                    "brand_color_light": brand_light,
                    "brand_color_dark":  brand_dark,
                })
            except Exception as e:
                print(f"[hero] Error processing {app_id}: {e}")

        return results

    def get_app_hero_image(self, name: str, pkg_type: str) -> str:
        """Return the first SOURCE screenshot URL for any package, or empty string.

        Strategy:
        - flatpak: name IS the AppStream component id  (e.g. com.discordapp.Discord)
        - pacman / aur: search pool by pkgname, then fall back to id search
        """
        IMAGE_KIND_SOURCE = AppStream.ImageKind.SOURCE

        def _first_screenshot(comp) -> str:
            for sc in comp.get_screenshots_all():
                for img in sc.get_images_all():
                    if img.get_kind() == IMAGE_KIND_SOURCE:
                        url = img.get_url() or ""
                        if url:
                            return url
            return ""

        try:
            if pkg_type == "flatpak":
                comp_box = self.pool.get_components_by_id(name)
                comps = comp_box.as_array() if hasattr(comp_box, 'as_array') else list(comp_box)
                if comps:
                    return _first_screenshot(comps[0])
            else:
                # pacman / aur: search by pkgname using lazy map
                if self._pkgname_map is None:
                    self._pkgname_map = {}
                    comps = self.pool.get_components()
                    arr = comps.as_array() if hasattr(comps, 'as_array') else list(comps)
                    for c in arr:
                        pkgname = c.get_pkgname()
                        if pkgname:
                            self._pkgname_map[pkgname] = c

                comp = self._pkgname_map.get(name)
                if comp:
                    return _first_screenshot(comp)

                # fallback: try name as an app-id
                comp_box = self.pool.get_components_by_id(name)
                comps = comp_box.as_array() if hasattr(comp_box, 'as_array') else list(comp_box)
                if comps:
                    return _first_screenshot(comps[0])
        except Exception as e:
            print(f"[hero_image] Error for {pkg_type}/{name}: {e}")

        return ""

File: src-rust/src/data/appstream_repository.rs
use std::collections::HashMap;
use std::path::PathBuf;
use async_trait::async_trait;
use crate::domain::entities::{HeroApp, Package};
use crate::domain::repositories::FeaturedRepository;
use crate::data::alpm_repository::AlpmPackageRepository;
use crate::data::flatpak_repository::FlatpakPackageRepository;

/// Calls `appstreamcli dump <app_id>` to get raw XML metadata.
fn appstreamcli_dump(app_id: &str) -> Option<String> {
    let out = std::process::Command::new("appstreamcli")
        .args(["dump", app_id])
        .output()
        .ok()?;
    if !out.status.success() { return None; }
    Some(String::from_utf8_lossy(&out.stdout).to_string())
}

/// Generic helper to extract the contents of simple tags (optionally with attributes)
fn extract_tag_with_attrs(xml: &str, tag: &str) -> String {
    let tag_open = format!("<{tag}");
    if let Some(start) = xml.find(&tag_open) {
        let abs_start = start;
        if let Some(tag_end) = xml[abs_start..].find('>') {
            let val_start = abs_start + tag_end + 1;
            let tag_close = format!("</{tag}>");
            if let Some(end) = xml[val_start..].find(&tag_close) {
                return xml[val_start..val_start + end].trim().to_string();
            }
        }
    }
    String::new()
}

/// Extract all `<icon>` tags and identify their type (stock, cached, remote) and value
fn extract_icons(xml: &str) -> Vec<(String, String)> {
    let mut icons = Vec::new();
    let mut pos = 0;
    while let Some(start) = xml[pos..].find("<icon") {
        let abs_start = pos + start;
        if let Some(tag_end) = xml[abs_start..].find('>') {
            let abs_tag_end = abs_start + tag_end;
            let tag_attrs = &xml[abs_start..abs_tag_end];
            let icon_type = if tag_attrs.contains("type=\"stock\"") {
                "stock"
            } else if tag_attrs.contains("type=\"cached\"") {
                "cached"
            } else if tag_attrs.contains("type=\"remote\"") {
                "remote"
            } else {
                ""
            };
            if let Some(end_idx) = xml[abs_tag_end..].find("</icon>") {
                let abs_end_idx = abs_tag_end + end_idx;
                let val = xml[abs_tag_end + 1 .. abs_end_idx].trim().to_string();
                icons.push((icon_type.to_string(), val));
                pos = abs_end_idx + "</icon>".len();
            } else {
                pos = abs_tag_end + 1;
            }
        } else {
            pos = abs_start + 1;
        }
    }
    icons
}

/// Find the first `<image type="source">` tag and return its URL
fn extract_first_source_image(xml: &str) -> String {
    let mut pos = 0;
    while let Some(start) = xml[pos..].find("<image") {
        let abs_start = pos + start;
        if let Some(tag_end) = xml[abs_start..].find('>') {
            let abs_tag_end = abs_start + tag_end;
            let tag_attrs = &xml[abs_start..abs_tag_end];
            if tag_attrs.contains("type=\"source\"") {
                if let Some(end_idx) = xml[abs_tag_end..].find("</image>") {
                    let abs_end_idx = abs_tag_end + end_idx;
                    return xml[abs_tag_end + 1 .. abs_end_idx].trim().to_string();
                }
            }
            pos = abs_tag_end + 1;
        } else {
            pos = abs_start + 1;
        }
    }
    String::new()
}

/// Parse primary colors from the `<branding>` tag if available
fn extract_branding_colors(xml: &str, fallback_light: &str, fallback_dark: &str) -> (String, String) {
    let mut light = fallback_light.to_string();
    let mut dark = fallback_dark.to_string();
    let mut pos = 0;
    while let Some(start) = xml[pos..].find("<color") {
        let abs_start = pos + start;
        if let Some(tag_end) = xml[abs_start..].find('>') {
            let abs_tag_end = abs_start + tag_end;
            let tag_attrs = &xml[abs_start..abs_tag_end];
            if tag_attrs.contains("type=\"primary\"") {
                if let Some(end_idx) = xml[abs_tag_end..].find("</color>") {
                    let abs_end_idx = abs_tag_end + end_idx;
                    let val = xml[abs_tag_end + 1 .. abs_end_idx].trim().to_string();
                    if !val.is_empty() {
                        if tag_attrs.contains("scheme_preference=\"light\"") {
                            light = val;
                        } else if tag_attrs.contains("scheme_preference=\"dark\"") {
                            dark = val;
                        }
                    }
                    pos = abs_end_idx + "</color>".len();
                    continue;
                }
            }
            pos = abs_tag_end + 1;
        } else {
            pos = abs_start + 1;
        }
    }
    (light, dark)
}

/// Searches `appstreamcli search <pkg_name>` to find the correct `app_id` (Identificador)
/// for a given pacman package name (`Pacote`).
fn find_app_id_by_pkgname(pkg_name: &str) -> Option<String> {
    let out = std::process::Command::new("appstreamcli")
        .args(["search", pkg_name])
        .output()
        .ok()?;
    if !out.status.success() { return None; }
    let text = String::from_utf8_lossy(&out.stdout);
    for block in text.split("---") {
        let mut current_id = None;
        let mut matches_pkg = false;
        for line in block.lines() {
            let line = line.trim();
            if let Some((key, val)) = line.split_once(':') {
                let key = key.trim();
                let val = val.trim();
                if key == "Identificador" {
                    let id = val.split_whitespace().next().unwrap_or(val);
                    current_id = Some(id.to_string());
                } else if key == "Pacote" {
                    if val == pkg_name || val.split(':').last() == Some(pkg_name) {
                        matches_pkg = true;
                    }
                }
            }
        }
        if matches_pkg {
            if let Some(id) = current_id {
                return Some(id);
            }
        }
    }
    None
}

pub struct AppStreamFeaturedRepository {
    alpm_repo: std::sync::Arc<AlpmPackageRepository>,
    #[allow(dead_code)]
    flatpak_repo: std::sync::Arc<FlatpakPackageRepository>,
    data_dir: PathBuf,
}

impl AppStreamFeaturedRepository {
    pub fn new(
        alpm_repo: std::sync::Arc<AlpmPackageRepository>,
        flatpak_repo: std::sync::Arc<FlatpakPackageRepository>,
    ) -> Self {
        // data/ directory lives next to the binary (or relative to workspace root)
        let data_dir = std::env::current_exe()
            .ok()
            .and_then(|p| p.parent().map(|p| p.to_path_buf()))
            .unwrap_or_else(|| PathBuf::from("."))
            .join("data");
        Self { alpm_repo, flatpak_repo, data_dir }
    }

    pub fn set_data_dir(mut self, path: PathBuf) -> Self {
        self.data_dir = path;
        self
    }

    fn load_app_ids(&self, filename: &str, fallback: &[&str]) -> Vec<String> {
        let path = self.data_dir.join(filename);
        if path.exists() {
            if let Ok(text) = std::fs::read_to_string(&path) {
                if let Ok(ids) = serde_json::from_str::<Vec<String>>(&text) {
                    return ids;
                }
            }
        }
        fallback.iter().map(|s| s.to_string()).collect()
    }

    async fn resolve_package(&self, app_id: &str) -> Option<Package> {
        let app_id = app_id.to_string();
        let alpm = self.alpm_repo.clone();

        tokio::task::spawn_blocking(move || {
            if let Some(xml) = appstreamcli_dump(&app_id) {
                let pkg_name = extract_tag_with_attrs(&xml, "pkgname");
                let pkg_name_opt = if pkg_name.is_empty() { None } else { Some(pkg_name) };

                let title = {
                    let t = extract_tag_with_attrs(&xml, "name");
                    if t.is_empty() { app_id.clone() } else { t }
                };
                let desc = extract_tag_with_attrs(&xml, "summary");

                let icons = extract_icons(&xml);
                let mut icon = String::new();
                if let Some((_, val)) = icons.iter().find(|(t, _)| t == "stock") {
                    icon = val.clone();
                } else if let Some((_, val)) = icons.first() {
                    icon = val.clone();
                }
                if icon.ends_with(".png") || icon.ends_with(".svg") || icon.ends_with(".xpm") {
                    if let Some((stem, _)) = icon.rsplit_once('.') {
                        icon = stem.to_string();
                    }
                }
                if icon.is_empty() {
                    icon = pkg_name_opt.clone().unwrap_or_else(|| app_id.clone());
                }

                if let Some(pkg_name) = pkg_name_opt {
                    // pacman package
                    let installed = alpm.is_installed_sync(&pkg_name);
                    let installed_version = alpm.get_installed_version_sync(&pkg_name);
                    let version = if !installed_version.is_empty() {
                        installed_version.clone()
                    } else {
                        get_sync_version(&pkg_name).unwrap_or_else(|| "latest".into())
                    };
                    Some(Package {
                        name: pkg_name,
                        title,
                        desc,
                        version,
                        pkg_type: "pacman".to_string(),
                        installed,
                        installed_version,
                        icon,
                    })
                } else {
                    // flatpak component
                    let fp_id = {
                        let id = extract_tag_with_attrs(&xml, "id");
                        if id.is_empty() { app_id.clone() } else { id }
                    };
                    let installed_flatpaks = flatpak_list_sync();
                    let installed = installed_flatpaks.contains_key(&fp_id);
                    let installed_version = installed_flatpaks.get(&fp_id).cloned().unwrap_or_default();
                    let version = if !installed_version.is_empty() { installed_version.clone() } else { "latest".into() };
                    Some(Package {
                        name: fp_id,
                        title,
                        desc,
                        version,
                        pkg_type: "flatpak".to_string(),
                        installed,
                        installed_version,
                        icon,
                    })
                }
            } else {
                // Fallback: treat app_id as a pacman/AUR package name directly
                let installed = alpm.is_installed_sync(&app_id);
                let installed_version = alpm.get_installed_version_sync(&app_id);
                let version = if !installed_version.is_empty() {
                    installed_version.clone()
                } else {
                    get_sync_version(&app_id).unwrap_or_else(|| "latest".into())
                };

                let (pkg_type, desc) = if installed {
                    let is_official = check_official_sync(&app_id);
                    let t = if is_official { "pacman" } else { "aur" };
                    (t.to_string(), "Pacote de sistema instalado.".to_string())
                } else if version != "latest" {
                    ("pacman".to_string(), String::new())
                } else {
                    ("aur".to_string(), "Ferramenta/Biblioteca do sistema (AUR)".to_string())
                };

                let title = app_id
                    .replace("-bin", "")
                    .replace("-git", "")
                    .replace('-', " ")
                    .split_whitespace()
                    .map(capitalize)
                    .collect::<Vec<_>>()
                    .join(" ");

                Some(Package {
                    name: app_id.clone(),
                    title,
                    desc,
                    version,
                    pkg_type,
                    installed,
                    installed_version,
                    icon: "package".to_string(),
                })
            }
        })
        .await
        .unwrap_or(None)
    }

    async fn resolve_hero(&self, app_id: &str, fallback_light: &str, fallback_dark: &str) -> Option<HeroApp> {
        let app_id = app_id.to_string();
        let fl = fallback_light.to_string();
        let fd = fallback_dark.to_string();
        let alpm = self.alpm_repo.clone();

        tokio::task::spawn_blocking(move || {
            let xml = appstreamcli_dump(&app_id)?;

            let hero_image = extract_first_source_image(&xml);
            let (brand_light, brand_dark) = extract_branding_colors(&xml, &fl, &fd);

            let pkg_name = extract_tag_with_attrs(&xml, "pkgname");
            let pkg_name_opt = if pkg_name.is_empty() { None } else { Some(pkg_name) };

            let title = {
                let t = extract_tag_with_attrs(&xml, "name");
                if t.is_empty() { app_id.clone() } else { t }
            };
            let desc = extract_tag_with_attrs(&xml, "summary");

            let icons = extract_icons(&xml);
            let mut icon = String::new();
            if let Some((_, val)) = icons.iter().find(|(t, _)| t == "stock") {
                icon = val.clone();
            } else if let Some((_, val)) = icons.first() {
                icon = val.clone();
            }
            if icon.ends_with(".png") || icon.ends_with(".svg") || icon.ends_with(".xpm") {
                if let Some((stem, _)) = icon.rsplit_once('.') {
                    icon = stem.to_string();
                }
            }
            if icon.is_empty() {
                icon = app_id.clone();
            }

            let (name, pkg_type, installed, installed_version, version) = if let Some(pkg_name) = pkg_name_opt {
                let installed = alpm.is_installed_sync(&pkg_name);
                let iv = alpm.get_installed_version_sync(&pkg_name);
                let ver = if !iv.is_empty() { iv.clone() } else {
                    get_sync_version(&pkg_name).unwrap_or_else(|| "latest".into())
                };
                (pkg_name, "pacman".to_string(), installed, iv, ver)
            } else {
                let fp_id = {
                    let id = extract_tag_with_attrs(&xml, "id");
                    if id.is_empty() { app_id.clone() } else { id }
                };
                let installed_flatpaks = flatpak_list_sync();
                let installed = installed_flatpaks.contains_key(&fp_id);
                let iv = installed_flatpaks.get(&fp_id).cloned().unwrap_or_default();
                let ver = if !iv.is_empty() { iv.clone() } else { "latest".into() };
                (fp_id, "flatpak".to_string(), installed, iv, ver)
            };

            Some(HeroApp {
                name,
                title,
                desc,
                version,
                pkg_type,
                installed,
                installed_version,
                icon,
                hero_image,
                brand_color_light: brand_light,
                brand_color_dark: brand_dark,
            })
        })
        .await
        .unwrap_or(None)
    }

}

// ─── Blocking helpers (called inside spawn_blocking) ──────────────────────────

fn flatpak_list_sync() -> HashMap<String, String> {
    let out = std::process::Command::new("flatpak")
        .args(["list", "--app", "--columns=application,version"])
        .output();
    let mut map = HashMap::new();
    if let Ok(o) = out {
        for line in String::from_utf8_lossy(&o.stdout).lines() {
            let parts: Vec<&str> = line.splitn(2, '\t').collect();
            if parts.len() == 2 {
                map.insert(parts[0].trim().to_string(), parts[1].trim().to_string());
            }
        }
    }
    map
}

fn get_sync_version(pkg_name: &str) -> Option<String> {
    let handle = alpm::Alpm::new("/", "/var/lib/pacman").ok()?;
    for dbname in ["core", "extra", "multilib"] {
        let _ = handle.register_syncdb(dbname, alpm::SigLevel::DATABASE_OPTIONAL);
    }
    for db in handle.syncdbs() {
        if let Ok(pkg) = db.pkg(pkg_name) {
            return Some(pkg.version().to_string());
        }
    }
    None
}

fn check_official_sync(pkg_name: &str) -> bool {
    let Ok(handle) = alpm::Alpm::new("/", "/var/lib/pacman") else { return false; };
    for dbname in ["core", "extra", "multilib"] {
        let _ = handle.register_syncdb(dbname, alpm::SigLevel::DATABASE_OPTIONAL);
    }
    handle.syncdbs().into_iter().any(|db| db.pkg(pkg_name).is_ok())
}

fn capitalize(s: &str) -> String {
    let mut c = s.chars();
    match c.next() {
        None => String::new(),
        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
    }
}

// ─── FeaturedRepository impl ──────────────────────────────────────────────────

#[async_trait]
impl FeaturedRepository for AppStreamFeaturedRepository {
    async fn get_featured_packages(&self) -> Vec<Package> {
        let ids = self.load_app_ids("featured_apps.json", &[
            "org.mozilla.firefox", "com.visualstudio.code", "org.kde.kdenlive",
            "org.gimp.GIMP", "org.videolan.VLC", "org.inkscape.Inkscape",
            "org.blender.Blender", "com.valvesoftware.Steam",
            "org.libreoffice.LibreOffice", "org.telegram.desktop",
        ]);
        let futures: Vec<_> = ids.iter().map(|id| self.resolve_package(id)).collect();
        let resolved = futures_util::future::join_all(futures).await;
        resolved.into_iter().flatten().collect()
    }

    async fn get_gaming_packages(&self) -> Vec<Package> {
        let ids = self.load_app_ids("gaming_apps.json", &[
            "com.valvesoftware.Steam", "net.lutris.Lutris", "goverlay",
            "gamemode", "mangohud", "com.discordapp.Discord",
            "com.heroicgameslauncher.hgl", "hydra-games-launcher-bin",
            "wine", "winetricks", "protontricks", "net.davidotek.pupgui2",
        ]);
        let futures: Vec<_> = ids.iter().map(|id| self.resolve_package(id)).collect();
        let resolved = futures_util::future::join_all(futures).await;
        resolved.into_iter().flatten().collect()
    }

    async fn get_development_packages(&self) -> Vec<Package> {
        let ids = self.load_app_ids("development_apps.json", &[
            "code", "visual-studio-code-bin", "neovim", "dev.zed.Zed",
            "jetbrains-toolbox", "gnome-builder", "nodejs", "fnm", "npm",
            "yarn", "pnpm", "bun", "docker", "docker-compose", "podman",
            "io.podman_desktop.PodmanDesktop", "distrobox",
            "rest.insomnia.Insomnia", "com.getpostman.Postman",
            "rustup", "cargo", "go", "python", "git", "github-cli",
            "lazygit", "zellij",
        ]);
        let futures: Vec<_> = ids.iter().map(|id| self.resolve_package(id)).collect();
        let resolved = futures_util::future::join_all(futures).await;
        resolved.into_iter().flatten().collect()
    }

    async fn get_hero_apps(&self) -> Vec<HeroApp> {
        let hero_ids = [
            ("com.valvesoftware.Steam",     "#c6d0f5", "#1b2838"),
            ("org.libreoffice.LibreOffice", "#c6d0f5", "#18A303"),
            ("net.lutris.Lutris",           "#f4b8e4", "#cc5500"),
            ("com.heroicgameslauncher.hgl", "#ac9aac", "#1c202c"),
            ("com.visualstudio.code",       "#c6d0f5", "#0078d4"),
            ("org.gimp.GIMP",              "#f4b8e4", "#4e3e6e"),
        ];
        let futures: Vec<_> = hero_ids.iter().map(|(id, fl, fd)| self.resolve_hero(id, fl, fd)).collect();
        let resolved = futures_util::future::join_all(futures).await;
        resolved.into_iter().flatten().collect()
    }

    async fn get_gaming_hero_apps(&self) -> Vec<HeroApp> {
        let hero_ids = [
            ("com.valvesoftware.Steam",     "#c6d0f5", "#1b2838"),
            ("net.lutris.Lutris",           "#f4b8e4", "#cc5500"),
            ("com.heroicgameslauncher.hgl", "#ac9aac", "#1c202c"),
        ];
        let futures: Vec<_> = hero_ids.iter().map(|(id, fl, fd)| self.resolve_hero(id, fl, fd)).collect();
        let resolved = futures_util::future::join_all(futures).await;
        resolved.into_iter().flatten().collect()
    }

    async fn get_development_hero_apps(&self) -> Vec<HeroApp> {
        let hero_ids = [
            ("com.visualstudio.code", "#c6d0f5", "#0078d4"),
            ("dev.zed.Zed",           "#f4b8e4", "#cc5500"),
            ("org.gnome.Builder",     "#cba6f7", "#1a1a24"),
        ];
        let futures: Vec<_> = hero_ids.iter().map(|(id, fl, fd)| self.resolve_hero(id, fl, fd)).collect();
        let resolved = futures_util::future::join_all(futures).await;
        resolved.into_iter().flatten().collect()
    }

    async fn get_app_hero_image(&self, name: &str, pkg_type: &str) -> String {
        let name = name.to_string();
        let pkg_type = pkg_type.to_string();

        tokio::task::spawn_blocking(move || {
            let app_id = if pkg_type == "flatpak" {
                name.clone()
            } else {
                if let Some(id) = find_app_id_by_pkgname(&name) {
                    id
                } else {
                    name.clone()
                }
            };

            if let Some(xml) = appstreamcli_dump(&app_id) {
                return extract_first_source_image(&xml);
            }
            String::new()
        })
        .await
        .unwrap_or_default()
    }
}
